### PR TITLE
--reportTimeToFile:{path to csv} flag for fsc.exe writing out collected activities

### PR DESCRIPTION
**Pull request by T-Gro**
This PR introduced a new **fsc.exe switch --reportTimeToFile: to write collected activities to a .csv** file.
This functionality is exposed in a way that other users of FCS can use too ( = not restricted to fsc.exe):
```
module Activity
      val addCsvFileListener: pathToFile: string -> IDisposable
```

To maintain a tabular structure of the .csv, only "well known" tag names are being exported.
This contains the main compilation phases (Parse > TypeCheck > ..) as well as gatherings for individual files.

As of now, **the old functionality of "--times" remains intact** (@vzarytovskii : Can we re-do what an existing flag  does completely?), which means the code is still there.

If we could redo --times behavior (can be a separate PR for sure), we could:
- Centralize 'CPU time' and 'GC collection' measurements to become an optional feature for any Activity
- Have the console and .csv exporters working with the same data.
- Add a filtering feature to enable smaller console output (? probably based on depth // distance from root activity? @safesparrow , @baronfel  - if you have some use cases in mind on what would be sensible to filter on, please comment here)

The functionality can be exercised for a full build process on a project incl. it's dependencies like this:
```
dotnet msbuild `
-v:n `
-t:Rebuild `
-p:Configuration=Debug `
-p:FscToolPath="C:\code\fsharp\artifacts\bin\fsc\Debug\net7.0" `
-p:FscToolExe="fsc.exe"  `
-p:AdditionalFscCmdFlags="--times --reportTimeToFile:C:\tests\bigTestViaActivitySource.csv"
```

Example of a produced file for multiple projects at once:
[bigTestViaActivitySource.csv](https://github.com/dotnet/fsharp/files/10187299/bigTestViaActivitySource.csv)

### CODE REVIEW

**Comment by vzarytovskii** [comment]
Can you please share an example of what does file look like?

**Comment by T-Gro** [comment]
> Yeah, I think it's fine to redo what the old flag does, as long as it will have the same "functionality" - will allow to easily look at what times are spent where.

That's good. It is not mentioned as part of --help and web documentation only says "Displays timing information for compilation.", which would be maintained. OK

**Comment by T-Gro** [comment]
> Can you please share an example of what does file look like?

Yep, uploaded.

File: src/Compiler/Driver/CompilerOptions.fs
Comment (KevinRansom) [inline src/Compiler/Driver/CompilerOptions.fs:2435]: commented code

File: src/Compiler/Checking/CheckExpressions.fs
Comment (T-Gro) [inline src/Compiler/Checking/CheckExpressions.fs:12098]: My bad - this is needed until my other PR about inlining is merged it, to prevent me from going crazy when using Debug version of FSC.

Comment (T-Gro) [inline src/Compiler/Checking/CheckExpressions.fs:12098]:
```suggestion
        assert(vspec.InlineInfo = inlineFlag)
```

Comment (KevinRansom) [inline src/Compiler/Checking/CheckExpressions.fs:12098]: commented code is not really a thing, unless it is documenting something about the surrounding code.

Comment (T-Gro) [reply src/Compiler/Checking/CheckExpressions.fs:12098]: It was just documenting that this failing assertion was driving me crazy, and that I want https://github.com/dotnet/fsharp/pull/14435 to get merged. 

Will clean up the commented places ofc 👍

File: src/Compiler/Driver/CompilerOptions.fs
Comment (T-Gro) [reply src/Compiler/Driver/CompilerOptions.fs:1746]: I intentionally did not want to conflict with an existing flag.
Not even sure if it is a good idea to have a flag that can be both on/off  as well one with a string argument.

Per Vlad's reply above, in a next RP I will unify the code with the old one (behind --times) to make them behave the same and change the appearance of --times slightly. Within this PR, --times was kept intact.

File: src/Compiler/Driver/CompilerConfig.fs
Comment (safesparrow) [inline src/Compiler/Driver/CompilerConfig.fs:520]: `Time` -> `Times`, or is there a reason behind the singular form?

File: src/Compiler/Driver/CompilerOptions.fs
Comment (safesparrow) [inline src/Compiler/Driver/CompilerOptions.fs:2405]: You could argue that it's possible to retain the old console formatting exactly as is, but only changing the underlying source of data - but maybe that's unnecessary work.

File: src/Compiler/Service/service.fs
Comment (safesparrow) [inline src/Compiler/Service/service.fs:688]: We have these/similar tag names in many more places - are you planning to unify them all in this PR, or was this just an example of what's to come?

File: src/Compiler/Driver/CompilerOptions.fs
Comment (T-Gro) [reply src/Compiler/Driver/CompilerOptions.fs:2405]: Yep, exactly - it is possible, but would mean dropping a more generalized approach and hardcoding the layout to fit what use to be there.

File: src/Compiler/Driver/CompilerConfig.fs
Comment (T-Gro) [reply src/Compiler/Driver/CompilerConfig.fs:520]: Not really.
Especially waiting on proposals for the cmd switch itself, since this is what is visible. Then I will align the names in code with it.

File: src/Compiler/Service/service.fs
Comment (T-Gro) [reply src/Compiler/Service/service.fs:688]: Will unify to make sure we use them, yes.

File: src/Compiler/Utilities/Activity.fs
Comment (vzarytovskii) [inline src/Compiler/Utilities/Activity.fs:124]: Can we use backgroundTask/task CE here?

File: src/Compiler/Driver/CompilerOptions.fs
Comment (KevinRansom) [reply src/Compiler/Driver/CompilerOptions.fs:1746]: We already do, as does C#:

````
--warnaserror[+|-]                       Report all warnings as errors
--warnaserror[+|-]:<warn;...>            Report specific warnings as errors
-
````
so:
--warnaserror
--warnaserror+
--warnaserror-
--warnaserror:42;35;0059
are all valid.

It's not just us so does C#:
````
                       - ERRORS AND WARNINGS -
-warnaserror[+|-]             Report all warnings as errors
-warnaserror[+|-]:<warn list> Report specific warnings as errors
                              (use "nullable" for all nullability warnings)
````

File: src/Compiler/Utilities/Activity.fs
Comment (T-Gro) [reply src/Compiler/Utilities/Activity.fs:124]: It becomes this. I do not see it as a big win TBH (when it comes to communicating intent, they feel rather the same to me)

```
            let writerTask = backgroundTask {
                use sw = new StreamWriter(path = pathToFile, append = true)
                for msg in messages.GetConsumingEnumerable() do
                    sw.WriteLine(msg)
                } 
```

File: src/Compiler/Driver/CompilerOptions.fs
Comment (T-Gro) [reply src/Compiler/Driver/CompilerOptions.fs:1746]: Good point, thanks

File: src/Compiler/Utilities/Activity.fs
Comment (vzarytovskii) [reply src/Compiler/Utilities/Activity.fs:124]: It's just a matter of us eating our own dogfood. I.e. using our corelib's features where suits.

Comment (T-Gro) [reply src/Compiler/Utilities/Activity.fs:124]: Ok, that explanation makes sense. Will change and retest in a concurrent scenario.

File: src/Compiler/Driver/CompilerOptions.fs
Comment (T-Gro) [reply src/Compiler/Driver/CompilerOptions.fs]: redone like this, there is a single flag called "times now"

--times   -> console only
--times:<file>

File: src/Compiler/Service/IncrementalBuild.fs
Comment (0101) [inline src/Compiler/Service/IncrementalBuild.fs:132]: Should these also go to `Activity.Tags`?

**Review by 0101 — DISMISSED** [review]

**Review by abonie — DISMISSED** [review]

File: src/Compiler/Utilities/Activity.fs
Comment (psfinaki) [inline src/Compiler/Utilities/Activity.fs:25]: Thanks for doing all this. I'd expect constants to be PascalCase but that's not very important.

File: src/Compiler/Driver/CompilerOptions.fs
Comment (psfinaki) [inline src/Compiler/Driver/CompilerOptions.fs:2441]: Sooo why this change?

File: src/Compiler/Utilities/Activity.fs
Comment (0101) [inline src/Compiler/Utilities/Activity.fs:86]: Quite surprisingly
```fsharp
> TimeSpan.FromSeconds(1).TotalSeconds.ToString("000.0000");;
val it: string = "001,0000"
```
This messes up the CSV.

File: src/Compiler/Utilities/Activity.fs
Comment (psfinaki) [inline src/Compiler/Utilities/Activity.fs:90]: Isn't Acitivity.RootId what we already need?
https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Activity.cs,284

File: src/Compiler/Utilities/Activity.fs
Comment (0101) [reply src/Compiler/Utilities/Activity.fs:86]:
```suggestion
        appendWithLeadingComma (a.Duration.TotalSeconds.ToString("000.0000", System.Globalization.CultureInfo.InvariantCulture))
```

File: src/Compiler/Service/IncrementalBuild.fs
Comment (T-Gro) [reply src/Compiler/Service/IncrementalBuild.fs:132]: They were only used at one place, they do not meet the "common enough" bar to make it.

File: src/Compiler/Driver/CompilerOptions.fs
Comment (T-Gro) [reply src/Compiler/Driver/CompilerOptions.fs:2441]: The activity must be .Dispose()'d only after the tags are enhanced (see above), not before.
But anyway, I have code lined up that removes this full function and reuses the generic Activity mechanism (a dedicated listeners for console logging).

File: src/Compiler/Utilities/Activity.fs
Comment (T-Gro) [reply src/Compiler/Utilities/Activity.fs:90]: I don't like what it does with string-based navigation https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Activity.cs,775d64ba2a5d0d39 , feels strange and allocate-y, compared to following one or few reference pointers.

File: src/Compiler/Utilities/Activity.fs
Comment (T-Gro) [reply src/Compiler/Utilities/Activity.fs:25]: There is an unrelated section about literals using the Literal attribute. Than PascalCase is expected.
But all the other let .. bindings of values I have seen around have been camelCase.

Comment (psfinaki) [reply src/Compiler/Utilities/Activity.fs:25]: Mhmhm yeah I probably mentally mixed it up with literals, nevermind then

File: src/Compiler/Utilities/Activity.fs
Comment (psfinaki) [reply src/Compiler/Utilities/Activity.fs:90]: I leave it to you. It looks like the impl above deals with some edge cases which we might or might not want to take into consideration

**Review by psfinaki — APPROVED** [review]

**Review by 0101 — APPROVED** [review]

**Pull request merged**
